README.md
Introduction to the mekko package
================

### Bar mekko graphics in R

This is a quick introduction to the bar mekko function\* in the mekko package.

Its main value is to add quantitative context to a bar graph, via bar width.

<br> Install with:

``` r
install.packages("mekko")
```

You can also install the development version with `devtools::install_github('ryninho/mekko', build_vignettes = TRUE)`.

View examples with:

``` r
vignette("mekko-vignette")
```

First we'll load ggplot and mekko and create some example data.

``` r
library(ggplot2)
library(mekko)
```

``` r
profits <- data.frame(
  product = c('widgets', 'sprockets', 'cogs', 'whosits', 'whatsits'),
  profit_margin = c(3.2, -1.4, 0.1, -5.5, 11.9),
  revenue = c(850, 1600, 900, 675, 250)
  )
```

### Bar mekko

Now let's take a look at profit margin by product using ggplot2.

``` r
ggplot(profits, aes(x = product, y = profit_margin)) +
  geom_bar(stat = "identity")
```

![](vignettes/mekko-vignette_files/figures/current-state-bad-example-bar-mekko-1.png)

Well that's insightful, but I don't know how worried I should be about the margin on whosits or cogs, nor do I know how happy I should be about whatsits knocking it out of the park. Maybe I can add revenue as the bar width so I know what's important here?

``` r
ggplot(profits, aes(x = product, y = profit_margin, width = revenue)) +
  geom_bar(stat = "identity") +
  labs(title = "Variable bar width fail :(")
```

    ## Warning: position_stack requires non-overlapping x intervals

    ## Warning: position_stack requires non-overlapping x intervals

![](vignettes/mekko-vignette_files/figures/ggplot-bar-width-fail-1.png)

Well shucks, that looks like some kind of Atari graphics airplane. Let's use the mekko package to put our margins in context.

``` r
bmx <- barmekko(profits, product, profit_margin, revenue)
bmx
```

![](vignettes/mekko-vignette_files/figures/bar-mekko-example-1.png)

Alright, so actually the weak profit margins on sprockets may be worth nearly as much focus as the problem with whosits. Also, no high-fives for margins on the whatsits until we triple sales of them.

Those labels are a little close together- this is a ggplot object so let's use the usual method of rotating the axes.

``` r
bmx + theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

![](vignettes/mekko-vignette_files/figures/bar-mekko-extension-1.png)

### Appendix

-   This package originally included a marimekko function as well, but the latter was discovered to be well covered in [other](https://CRAN.R-project.org/package=ggmosaic/vignettes/ggmosaic.html) [packages](https://www.semanticscholar.org/paper/Product-Plots-Wickham-Hofmann/0598a59354cb96161d68dab91fb0de21fb8671fd/figure/6), so it was removed to simplify maintenance
-   While the barmekko is a ggplot object, your mileage may vary when attempting operations that normally would work on a ggplot object, as the data frame is deconstructed in the process of creating the graph


NEWS.md
# mypackage 0.1.0

First submission.


examples.R
library(ggplot2)

#' Display square for a given centroid and edge measure.
#'
#'Square geom is displayed on the graph
#'
#' \code{geom_draw_sqr} uses the center of the tile and its size (\code{x},
#' \code{y}, \code{edge}). 
#' 
#' @param mapping Set of aesthetic mappings created
#'  by aes or aes_. If specified and inherit.aes = TRUE
#'  (the default), is combined with the default mapping at
#'  the top level of the plot. You only need to supply
#'  mapping if there isn't a mapping defined for the plot.
#'
#' @param data A data frame. If specified, overrides the default
#'  data frame defined at the top level of the plot.
#'
#' @param stat The statistical transformation to use on
#'  the data for this layer, as a string.
#'
#' @param position Position adjustment, either as a string,
#'  or the result of a call to a position adjustment function.
#'
#' @param na.rm FALSE (the default), removes missing values
#'  with a warning. If TRUE silently removes missing values.
#'
#' @param show.legend Logical. Should this layer be included in
#'  the legends? NA, the default, includes if any aesthetics are
#'  mapped. FALSE never includes, and TRUE always includes.
#'
#' @param inherit.aes If FALSE, overrides the default aesthetics,
#'  rather than combining with them. This is most useful for helper
#'  functions that define both data and aesthetics and shouldn't inherit
#'  behaviour from the default plot specification
#'
#' @param ... Other arguments passed on to layer. There are three types of arguments you can use here.
#'
#' @section Aesthetics:
#' 
#' @examples
#' # Draw squares of varied sizes at aribtary locations
#'  df = data.frame(x = c(10,20,30,40,50),
#'                  y = c(50,40,30,20,10),
#'                  edge = (10,15,20,25,30),
#'                  type = c("A","B","A","A","B"))
#'  ggplot(df,aes(x = x,y = y,edge = edge)) + geom_draw_sqr(aes(fill = type))
#' @export


geom_draw_sqr = function(mapping = NULL, data = NULL, stat = "identity",
                         position = "identity", na.rm = FALSE, show.legend = NA, 
                         inherit.aes = TRUE, ...) {
  layer(
    geom = GeomDrawSquare, mapping = mapping, data = data, stat = stat, 
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(na.rm = na.rm, ...)
  )
}

#'  @export   


GeomDrawSquare =  ggproto("GeomDrawSquare", Geom,
                          required_aes = c("x","y","edge"),
                          default_aes = aes(
                            colour = "white", fill = "grey20", size = 0.5,
                            linetype = 1, alpha = 1
                          ),
                          draw_key = draw_key_rect,
                          draw_group = function(data,panel_scales,coord)
                          {
                            
                            x = c((data$x -data$edge/2),(data$x+data$edge/2))
                            y = c((data$y -data$edge/2),(data$y+data$edge/2))
                            data2 = data.frame(x = x, y = y)
                            attr = data[1,,drop = F]
                            coords = coord$transform(data2,panel_scales)
                            sd = abs(coords$x[1:nrow(data)] - coords$x[(nrow(data)+1):(nrow(data)*2)])
                            grid::rectGrob(x = (coords$x[1:nrow(data)]+sd/2), 
                                           y = (coords$y[1:nrow(data)]+sd/2),
                                           width = sd,height = sd,  default.units = "native",
                                           gp = grid::gpar(
                                             col = attr$colour,
                                             fill = scales::alpha(attr$fill,attr$alpha),
                                             lwd = attr$size,
                                             lty = attr$linetype
                                           ))
                          })

df = data.frame(x = c(10,20,30,40,50),
                y = c(50,40,30,20,10),
                edge = c(10,15,20,25,30),
                type = c("A","B","A","A","B"))
ggplot(df,aes(x = x,y = y,edge = edge)) + geom_draw_sqr(aes(fill = type))


# https://raw.githubusercontent.com/heike/mosaic/master/hard/R/geomtilenew.R

#' Draw rectangles with wide white borders.
#'
#' \code{geom_tile} uses the center of the tile and its size (\code{x},
#' \code{y}, \code{width}, \code{height}). 
#' 
#' @param mapping Set of aesthetic mappings created
#'  by aes or aes_. If specified and inherit.aes = TRUE
#'  (the default), is combined with the default mapping at
#'  the top level of the plot. You only need to supply
#'  mapping if there isn't a mapping defined for the plot.
#'
#' @param data A data frame. If specified, overrides the default
#'  data frame defined at the top level of the plot.
#'
#' @param stat The statistical transformation to use on
#'  the data for this layer, as a string.
#'
#' @param position Position adjustment, either as a string,
#'  or the result of a call to a position adjustment function.
#'
#' @param na.rm f FALSE (the default), removes missing values
#'  with a warning. If TRUE silently removes missing values.
#'
#' @param show.legend logical. Should this layer be included in
#'  the legends? NA, the default, includes if any aesthetics are
#'  mapped. FALSE never includes, and TRUE always includes.
#'
#' @param inherit.aes If FALSE, overrides the default aesthetics,
#'  rather than combining with them. This is most useful for helper
#'  functions that define both data and aesthetics and shouldn't inherit
#'  behaviour from the default plot specification
#'
#' @param ... other arguments passed on to layer. There are three types of arguments you can use here.
#'
#' @section Aesthetics:
#' 
#' @examples
#' # To draw arbitrary rectangles with wide white borders, use geom_tile_new()
#' df <- data.frame(
#'   x = rep(c(2, 5, 7, 9, 12), 2),
#'   y = rep(c(1, 2), each = 5),
#'   z = factor(rep(1:5, each = 2)),
#'   w = rep(diff(c(0, 4, 6, 8, 10, 14)), 2)
#' )
#' ggplot(df, aes(x, y)) +
#'   geom_tile_new(aes(fill = z))
#' 
#' @export

geom_tile_new <- function(mapping = NULL, data = NULL,
                          stat = "identity", position = "identity",
                          ...,
                          na.rm = FALSE,
                          show.legend = NA,
                          inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomTileNew,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      ...
    )
  )
}


#' @export

GeomTileNew <- ggproto("GeomTileNew", GeomRect,
                       required_aes = c("x", "y"),
                       default_aes = aes(fill = "grey20", colour = "ehite", size = 2, linetype = 1, alpha = NA),
                       draw_key = draw_key_polygon,
                       
                       extra_params = c("na.rm", "width", "height"),
                       
                       setup_data = function(data, params) {
                         data$width <- data$width %||% params$width %||% resolution(data$x, FALSE)
                         data$height <- data$height %||% params$height %||% resolution(data$y, FALSE)
                         
                         transform(data,
                                   xmin = x - width / 2,  xmax = x + width / 2,  width = NULL,
                                   ymin = y - height / 2, ymax = y + height / 2, height = NULL
                         )
                       }
)



# EXAMPLE of geom_tile_new ----------------------------------------------------------------------------
"%||%" <- function(a, b) {
  if (!is.null(a)) a else b
}
# If you want to draw arbitrary rectangles, use geom_tile() or geom_rect()
df <- data.frame(
  x = rep(c(2, 5, 7, 9, 12), 2),
  y = rep(c(1, 2), each = 5),
  z = factor(rep(1:5, each = 2)),
  w = rep(diff(c(0, 4, 6, 8, 10, 14)), 2)
)
ggplot(df, aes(x, y)) +
  geom_tile_new(aes(fill = z, width = w))



scratchpad.R
library(ggplot2) # TODO: is this the right place for this?
library(dplyr)

hec <- HairEyeColor %>% data.frame %>% filter(Sex == "Male") %>% select(-Sex)

# standard example
ggplot(hec, aes(x = Eye, y = Freq, fill = Hair)) + 
  geom_bar(stat = "identity", position = "fill")

# ... can use width, but...
ggplot(hec, aes(x = Eye, y = Freq, fill = Hair, 
  width = c(.1, .1, .1, .1, .6, .6, .6, .6, 1, 1, 1, 1, .5, .5, .5, .5))) + 
  geom_bar(stat = "identity", position = "fill")

# ... only with a few categories and nothing too large!
ggplot(hec, aes(x = Eye, y = Freq, fill = Hair, 
  width = c(.1, .1, .1, .1, 6, 6, 6, 6, 1, 1, 1, 1, .5, .5, .5, .5))) + 
  geom_bar(stat = "identity", position = "fill")

# ... and fails if trying to place the labels since axis is discrete
ggplot(hec, aes(x = Eye, y = Freq, fill = Hair, 
  width = c(.1, .1, .1, .1, 6, 6, 6, 6, 1, 1, 1, 1, .5, .5, .5, .5))) + 
  geom_bar(stat = "identity", position = "fill") + 
  scale_x_continuous(breaks = c(.1, 6.1, 7.1, 7.6), 
                     labels = c("Brown", "Blue", "Hazel", "Green"))



# example -----------------------------------------------------------------


df <- hec %>% rename(x = Eye, y = Hair, width = Freq)

xlabel <- substitute(x)
ylabel <- substitute(y)

x_widths <- df %>% group_by(x) %>% 
  summarize(x_width = sum(width)) %>% 
  mutate(
    wmin = c(0, head(cumsum(x_width), length(x_width) - 1)),
    wmax = cumsum(x_width),
    wcenter = (wmin + wmax) / 2
  )

df <- df %>% inner_join(x_widths, by = "x")

df <- df %>% group_by(x) %>% 
  do(mutate(., 
    ymin = c(0, head(cumsum(width), length(width) - 1)),
    ymax = cumsum(width),
    ymin = ymin / max(ymax),
    ymax = ymax / max(ymax))) %>%
  ungroup
arrange(df, x, y)

p <- ggplot(df, aes(xmin=wmin, xmax=wmax, ymin=ymin, ymax=ymax, fill = y))

p <- p + geom_rect()

breaks <- unique(df$wcenter)
labels <- unique(df$y)
p <- p + scale_x_continuous(breaks = breaks, labels = labels)

p + xlab(xlabel) + ylab(ylabel) + guides(fill=guide_legend(title=xlabel))


# function definition -----------------------------------------------------

marimekko <- function(df, x, y, width) {
  xlabel <- substitute(x)
  ylabel <- substitute(y)
  df$x <- eval(substitute(x), df)
  df$y <- eval(substitute(y), df)
  df$width <- eval(substitute(width), df)
  
  x_widths <- df %>% group_by(x) %>% 
    summarize(x_width = sum(width)) %>% 
    mutate(
      wmin = c(0, head(cumsum(x_width), length(x_width) - 1)),
      wmax = cumsum(x_width),
      wcenter = (wmin + wmax) / 2
    )
  
  df <- df %>% inner_join(x_widths, by = "x")
  
  df <- df %>% group_by(x) %>% 
    do(mutate(., 
              ymin = c(0, head(cumsum(width), length(width) - 1)),
              ymax = cumsum(width),
              ymin = ymin / max(ymax),
              ymax = ymax / max(ymax))) %>%
    ungroup
  arrange(df, x, y)
  
  p <- ggplot(df, aes(xmin=wmin, xmax=wmax, ymin=ymin, ymax=ymax, fill = y))
  
  p <- p + geom_rect()
  
  breaks <- unique(df$wcenter)
  labels <- unique(df$y)
  p <- p + scale_x_continuous(breaks = breaks, labels = labels)
  
  p + xlab(xlabel) + ylab(ylabel) + guides(fill=guide_legend(title=xlabel)) 
}


# actual function call ----------------------------------------------------

marimekko(hec, Eye, Hair, Freq)


R/marimekko.R
#' Create a Marimekko plot.
#'
#' A smarter stacked bar chart.
#'
#' @param data A data frame.
#' @param x A categorical variable defining the width categories.
#' @param y A categorical variable defining the (vertical) segment categories.
#' @param width A numeric variable to be summed across both categories.
#' @return A Marimekko constructed with ggplot2.
#' @importFrom dplyr %>%
#' @importFrom dplyr ungroup
#' @export
#' @examples
#' library(ggplot2)
#' hec <- dplyr::filter(data.frame(HairEyeColor), Sex == "Male")
#' marimekko(hec, Eye, Hair, Freq)
#' # Note: Compare to productplots::prodplot(
#' #   hec, Freq ~ Hair + Eye, mosaic("v")
#' # ) + aes(fill=Hair)
#' marimekko(hec, Hair, Eye, Freq) + labs(title = "Hair and Eye Color")
marimekko <- function(data, x, y, width) {
  df <- data
  xlabel <- as.character(substitute(x))
  ylabel <- as.character(substitute(y))
  df$x <- eval(substitute(x), df) %>% as.character
  df$y <- eval(substitute(y), df)
  df$width <- eval(substitute(width), df)

  x_widths <- df %>% dplyr::group_by(x) %>%
    dplyr::summarize(x_width = sum(width)) %>%
    dplyr::mutate(
      wmin = c(0, head(cumsum(x_width), length(x_width) - 1)),
      wmax = cumsum(x_width),
      wcenter = (wmin + wmax) / 2
    )

  df <- df %>% dplyr::inner_join(x_widths, by = "x")

  df <- df %>% dplyr::group_by(x) %>%
    dplyr::do(dplyr::mutate(.,
      ymin = c(0, head(cumsum(width), length(width) - 1)),
      ymax = cumsum(width),
      ymin = ymin / max(ymax),
      ymax = ymax / max(ymax))) %>%
    ungroup

  p <- ggplot2::ggplot(df, ggplot2::aes(
         xmin = wmin, xmax = wmax, ymin = ymin, ymax = ymax, fill = y
         )
    )

  p <- p + ggplot2::geom_rect(color = "gray33")

  breaks <- unique(df$wcenter)
  labels <- unique(df$x) %>% as.character
  p <- p + ggplot2::scale_x_continuous(breaks = breaks, labels = labels)

  p + ggplot2::xlab(xlabel) + ggplot2::ylab(ylabel) +
    ggplot2::guides(fill = ggplot2::guide_legend(title = ylabel))
}


tutorial_examples.R
StatChull <- ggproto("StatChull", Stat,
                     compute_group = function(data, scales) {
                       data[chull(data$x, data$y), , drop = FALSE]
                     },
                     
                     required_aes = c("x", "y")
)


GeomPolygonHollow <- ggproto("GeomPolygonHollow", GeomPolygon,
                             default_aes = aes(colour = "black", fill = NA, size = 0.5, linetype = 1, 
                                               alpha = NA)
)

geom_chull <- function(mapping = NULL, data = NULL, 
                       position = "identity", na.rm = FALSE, show.legend = NA, 
                       inherit.aes = TRUE, ...) {
  layer(
    stat = StatChull, geom = GeomPolygonHollow, data = data, mapping = mapping,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(na.rm = na.rm, ...)
  )
}

ggplot(mpg, aes(displ, hwy)) + 
  geom_point() + 
  geom_chull()


ggplot(mpg, aes(displ, hwy)) + 
  geom_point() + 
  layer(
    stat = StatChull, geom = GeomPolygonHollow, data = NULL, mapping = NULL,
    position = "identity", show.legend = NA, inherit.aes = TRUE,
    params = list(na.rm = TRUE)
  )


stack_overflow_answers.R
makeplot_mosaic <- function(data, x, y, ...){
  xvar <- deparse(substitute(x))
  yvar <- deparse(substitute(y))
  mydata <- data[c(xvar, yvar)];
  mytable <- table(mydata);
  widths <- c(0, cumsum(apply(mytable, 1, sum)));
  heights <- apply(mytable, 1, function(x){c(0, cumsum(x/sum(x)))});
  
  alldata <- data.frame();
  allnames <- data.frame();
  for(i in 1:nrow(mytable)){
    for(j in 1:ncol(mytable)){
      alldata <- rbind(alldata, c(widths[i], widths[i+1], heights[j, i], heights[j+1, i]));
    }
  }
  colnames(alldata) <- c("xmin", "xmax", "ymin", "ymax")
  
  alldata[[xvar]] <- rep(dimnames(mytable)[[1]],rep(ncol(mytable), nrow(mytable)));
  alldata[[yvar]] <- rep(dimnames(mytable)[[2]],nrow(mytable));
  
  ggplot(alldata, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax)) + 
    geom_rect(color="black", aes_string(fill=yvar)) +
    xlab(paste(xvar, "(count)")) + ylab(paste(yvar, "(proportion)"));
}

makeplot_mosaic(mtcars, vs, gear)

library(plotly)
ggplotly()



ggMMplot <- function(var1, var2){
  require(ggplot2)
  levVar1 <- length(levels(var1))
  levVar2 <- length(levels(var2))
  
  jointTable <- prop.table(table(var1, var2))
  plotData <- as.data.frame(jointTable)
  plotData$marginVar1 <- prop.table(table(var1))
  plotData$var2Height <- plotData$Freq / plotData$marginVar1
  plotData$var1Center <- c(0, cumsum(plotData$marginVar1)[1:levVar1 -1]) +
    plotData$marginVar1 / 2
  
  ggplot(plotData, aes(var1Center, var2Height)) +
    geom_bar(stat = "identity", aes(width = marginVar1, fill = var2), col = "Black") +
    geom_text(aes(label = as.character(var1), x = var1Center, y = 1.05)) 
}

ggMMplot(diamonds$cut, diamonds$clarity)
ggplotly()


R/barmekko.R
#' Calculate positions from widths.
#'
#' @param width A numeric vector of bar widths.
#' @return A numeric vector of bar positions.
positions <- function(width) {
  0.5 * (cumsum(width) + cumsum(c(0, width[-length(width)])))
}

#' Create a bar mekko plot.
#'
#' A smarter bar chart.
#'
#' @param data A data frame.
#' @param x A categorical variable defining the width categories.
#' @param y A numeric variable defining the bar height.
#' @param width A numeric variable defining the bar widths
#' @param values A boolean indicating whether to show value labels in bars
#' @return A bar mekko constructed with ggplot2.
#' @export
#' @examples
#' library(ggplot2)
#' df <- data.frame(
#'   region = c('Northeast', 'Southeast', 'Central', 'West'),
#'   sales = c(1200, 800, 450, 900),
#'   avg_margin = c(3.2, -1.4, 0.1, 2.1)
#'   )
#' barmekko(df, region, avg_margin, sales)
#' barmekko(df, region, avg_margin, sales) + labs(title = 'Margins by Region')
#' barmekko(df[order(-df$sales),], region, avg_margin, sales)
#' barmekko(df[order(-df$avg_margin),], region, avg_margin, sales)
barmekko <- function(data, x, y, width, values = FALSE) {
  df <- data
  xlabel <- as.character(substitute(x))
  ylabel <- as.character(substitute(y))
  x <- as.character(eval(substitute(x), df))
  y <- eval(substitute(y), df)
  width <- eval(substitute(width), df)
  pos <- positions(width)
  p <- suppressWarnings(
    ggplot2::ggplot() +
    ggplot2::geom_bar(ggplot2::aes(x = pos, width = width, y = y, fill = x),
      stat = "identity") +
    ggplot2::scale_x_continuous(labels = x, breaks = pos) +
    ggplot2::xlab(xlabel) +
    ggplot2::ylab(ylabel) +
    ggplot2::guides(fill = ggplot2::guide_legend(title = xlabel))
  )
  if(values) {
    p + ggplot2::geom_text(ggplot2::aes(x = pos, y = 0, label = y, vjust = -0.5))
  } else {
    p
  }
}
